cli/flox-rust-sdk/src/models/environment/generations.rs

```diff
@@ -673,6 +928,361 @@ mod tests {
     const GEN_ID_1: GenerationId = GenerationId(1);
     const GEN_ID_2: GenerationId = GenerationId(2);
 
+    const AUTHOR: &str = "author";
+    const HOSTNAME: &str = "host";
+
+    fn default_add_generation_options() -> AddGenerationOptions {
+        AddGenerationOptions {
+            author: AUTHOR.into(),
+            hostname: HOSTNAME.into(),
+            timestamp: Utc::now(),
+            kind: HistoryKind::Other {
+                summary: "mock".into(),
+            },
+        }
+    }
+
+    fn default_switch_generation_options(next_generation: GenerationId) -> SwitchGenerationOptions {
+        SwitchGenerationOptions {
+            author: AUTHOR.into(),
+            hostname: HOSTNAME.into(),
+            timestamp: Utc::now(),
+            next_generation,
+        }
+    }
+
+    mod metadata {
+        use chrono::Utc;
+        use pretty_assertions::{assert_eq, assert_str_eq};
+        use serde_json::{Value, json};
+
+        use super::{AUTHOR, HOSTNAME, default_switch_generation_options};
+        use crate::models::environment::generations::tests::default_add_generation_options;
+        use crate::models::environment::generations::{
+            AllGenerationsMetadata,
+            GenerationId,
+            GenerationsError,
+            HistoryKind,
+            HistorySpec,
+            SwitchGenerationOptions,
+        };
+
+        /// Adding a generation adds consisten metadata, ie.
+        ///
+        /// * adds new [SingleGenerationMetadata]
+        /// * updates the current generation
+        /// * adds a history entry for adding the generation
+        #[test]
+        fn add_generation_adds_metadata_and_history() {
+            let mut metadata = AllGenerationsMetadata::default();
+
+            let options = default_add_generation_options();
+
+            let (generation, generation_metadata, history) =
+                metadata.add_generation(options.clone());
+
+            let (generation_metadata, history) = (generation_metadata.clone(), history.clone());
+
+            assert_eq!(metadata.current_gen, Some(generation));
+            assert_eq!(generation_metadata.created, options.timestamp);
+            assert_eq!(generation_metadata.last_active, Some(options.timestamp));
+
+            assert_eq!(history.author, options.author);
+            assert_eq!(history.hostname, options.hostname);
+            assert_eq!(history.current_generation, generation);
+            assert_eq!(history.previous_generation, None);
+            assert_eq!(history.info, options.kind);
+            assert_eq!(history.timestamp, options.timestamp);
```

Comment:
(non-blocking, same for `assert_switched_state` further down) Asserting on the entire object makes it easier to debug failures because you don't have to step through each field one-by-one and it prevents us from forgetting to add assertions on new fields in the future (requires an `Eq, PartialEq` on `SingleGenerationMetadata`):
```suggestion
            assert_eq!(generation_metadata, SingleGenerationMetadata {
                created: options.timestamp,
                last_active: Some(options.timestamp),
                description: history.summary(),
            });
            assert_eq!(history, HistorySpec {
                info: options.kind,
                author: options.author,
                hostname: options.hostname,
                timestamp: options.timestamp,
                current_generation: generation,
                previous_generation: None,
            });
```